### DISCUSSION TITLE
list: about add macro LIST_FOREACH_SAFE

### DISCUSSION THREAD

**Discussion by jobo-zt** [baresip/re#1200] [Q&A]
sreimers, Thank you for your reply.
1. The first approach directly modifies the le pointer during the list traversal. If an exception occurs or the list structure is modified in list_unlink or mem_deref, it may cause the traversal to be interrupted or access already freed memory.
2. The second approach uses the LIST_FOREACH_SAFE macro, which safely traverses the list using two pointers, le and n. Even if the list structure is modified in list_unlink or mem_deref, it will not affect the safety of the traversal.

I tried to use it, but it didn't work.
```c
#if 0	
	LIST_FOREACH_SAFE(&turndp()->rm_map, le, n)
	{
		struct udp_socks *uks = list_ledata(le);
		if (thrd_id == uks->thrd_id) {

			udp_thread_detach(uks->rel_us);
			udp_thread_detach(uks->rsv_us);

			list_unlink(&uks->le);
			mem_deref(uks);
		}
	}
#endif
#if 1
	le = list_head(&turndp()->rm_map);
	while (le)
	{
		struct udp_socks *uks = list_ledata(le);
		if (thrd_id == uks->thrd_id) {
			le = le->next;

			udp_thread_detach(uks->rel_us);
			udp_thread_detach(uks->rsv_us);
			
			list_unlink(&uks->le);
			mem_deref(uks);
		}
	}
#endif
```

**Comment by jobo-zt** [comment]
https://github.com/baresip/re/pull/1197

**Comment by sreimers** [comment]
Try this one:

```c
	le = list_head(&turndp()->rm_map);
	while (le)
	{
		struct udp_socks *uks = list_ledata(le);
                le = le->next;
		if (thrd_id == uks->thrd_id) {
			udp_thread_detach(uks->rel_us);
			udp_thread_detach(uks->rsv_us);
			
			list_unlink(&uks->le);
			mem_deref(uks);
		}
	}
```

> 1. The first approach directly modifies the `le` pointer during the list traversal. If an exception occurs or the list structure is modified in `list_unlink` or `mem_deref`, it may cause the traversal to be interrupted or access already freed memory.

The original pointer is not modified, only the reference pointer is set to the next element and should not be used after this (before next round).

**Comment by jobo-zt** [comment]
Thank you for your reply. I tried to use it. it is ok. I will submit a PR for the turn_thread.
